service/history/workflow/retry_test.go

```diff
@@ -153,241 +152,192 @@ func Test_IsRetryable(t *testing.T) {
-		ai.Attempt = 6
-		ai.RetryMaximumAttempts = 8
-		ai.RetryExpirationTime = timestamppb.New(now.Add(time.Millisecond * 5))
-		interval, retryState := getBackoffInterval(
-			now,
-			ai.Attempt,
-			ai.RetryMaximumAttempts,
-			ai.RetryInitialInterval,
-			ai.RetryMaximumInterval,
-			ai.RetryExpirationTime,
-			ai.RetryBackoffCoefficient,
-			serverFailure,
-			ai.RetryNonRetryableErrorTypes,
+	t.Run("when max attempts is set to 0 should keep trying", func(t *testing.T) {
+		initialDelay := 2 * time.Second
+		interval, retryState := nextBackoffInterval(
+			doNotCare(now),
+			10,
+			0,
+			doNotCare(initInterval(initialDelay)),
+			doNotCare(maxInterval(30*time.Minute)),
+			doNotCare(expirationIn(60*time.Minute)),
+			2,
 		)
-		a.Equal(backoff.NoBackoff, interval)
-		a.Equal(enumspb.RETRY_STATE_TIMEOUT, retryState)
+		assert.Equal(t, initialDelay*pow(2, 10-1), interval)
+		assert.Equal(t, enumspb.RETRY_STATE_IN_PROGRESS, retryState)
 	})
 
-	t.Run("extend expiration, next interval should be 10s", func(t *testing.T) {
-		ai.RetryExpirationTime = timestamppb.New(now.Add(time.Minute))
-		interval, retryState := getBackoffInterval(
-			now,
-			ai.Attempt,
-			ai.RetryMaximumAttempts,
-			ai.RetryInitialInterval,
-			ai.RetryMaximumInterval,
-			ai.RetryExpirationTime,
-			ai.RetryBackoffCoefficient,
-			serverFailure,
-			ai.RetryNonRetryableErrorTypes,
+	t.Run("if expiration is not 0 and expected delay beyond expiration should return no more retries", func(t *testing.T) {
+		initialDelay := 2 * time.Second
+		interval, retryState := nextBackoffInterval(
+			doNotCare(now),
+			10,
+			0,
+			initInterval(initialDelay),
+			maxInterval(30*time.Minute),
+			expirationIn(1*time.Minute),
+			2,
 		)
-		a.Equal(time.Millisecond*10, interval)
-		a.Equal(enumspb.RETRY_STATE_IN_PROGRESS, retryState)
+		assert.Equal(t, backoff.NoBackoff, interval)
+		assert.Equal(t, enumspb.RETRY_STATE_TIMEOUT, retryState)
 	})
 
-	t.Run("with big max retry, math.Pow() could overflow, verify that it uses the MaxInterval", func(t *testing.T) {
-		ai.Attempt = 64
-		ai.RetryMaximumAttempts = 100
-		interval, retryState := getBackoffInterval(
-			now,
-			ai.Attempt,
-			ai.RetryMaximumAttempts,
-			ai.RetryInitialInterval,
-			ai.RetryMaximumInterval,
-			ai.RetryExpirationTime,
-			ai.RetryBackoffCoefficient,
-			serverFailure,
-			ai.RetryNonRetryableErrorTypes,
+	t.Run("if expiration is 0 should retry", func(t *testing.T) {
+		initialDelay := 2 * time.Second
+		interval, retryState := nextBackoffInterval(
+			doNotCare(now),
+			10,
+			0,
+			initInterval(initialDelay),
+			maxInterval(30*time.Minute),
+			expirationIn(0),
+			2,
 		)
-		a.Equal(time.Millisecond*10, interval)
-		a.Equal(enumspb.RETRY_STATE_IN_PROGRESS, retryState)
+		assert.Equal(t, backoff.NoBackoff, interval)
+		assert.Equal(t, enumspb.RETRY_STATE_TIMEOUT, retryState)
 	})
 }
+
+func doNotCare[T any](x T) T { return x }
```

Comment:
At first I was a bit confused; but I really like this! If we were using a property-based testing tool, this could be a randomly generated value.

Reply:
`TestVars.Any()`
https://github.com/temporalio/temporal/blob/1be76e3583ef01ba9f79503e81fed5b7c9ab389c/tests/testvars/test_vars.go#L211

Reply:
It shouldn't be in functional tests package but in some common test helpers package.

Reply:
And it needs to have `Any` for every type. I am going to improve that `testvars` package and will use it here too. One day.

Reply:
Unfortunately, I can't use Any(). There are some invariants that should hold for the values which I marked with `doNotCare`.  At the same time, I want the reader to easily distinguish between values which can be changed without causing the test fail and the values which are important to the test success. Hope this makes sense.